File: AIiOS/AIiOS/Services/NeedleEmbedder.swift
import Foundation
import os

private let needleLog = Logger(subsystem: "ai.ecoinference.app", category: "NeedleEmbedder")

/// Cactus Needle 3 as a fixed feature extractor for the router's semantic
/// facts (docs/ROUTING.md §4 — adopted design). The model ships once; ALL
/// tunable intelligence (exemplar texts, thresholds, centering mean) lives in
/// `ExemplarConfig` JSON — bundled default plus live updates via Firebase
/// Remote Config key `router_exemplars`, same version-gated pattern as
/// `router_rules`. Mirrors Android NeedleEmbedder.kt.
///
/// Failure contract: every failure path leaves `semanticFacts` returning nil
/// and the router falls back to keyword facts. Routing must never block or
/// crash chat. Engine unavailable (weights not vendored, stub bridge, init
/// error) is a normal state, not an error.
///
/// Threading: the needle C API is one process-global, NOT thread-safe model.
/// Every native call runs on `queue`; callers may invoke `semanticFacts`
/// from any thread except the queue itself (it blocks on queue.sync).
final class NeedleEmbedder {

    static let shared = NeedleEmbedder()

    struct ExemplarConfig: Codable {
        let version: Int
        /// fact name -> exemplar texts (embedded once per config load).
        let categories: [String: [String]]
        /// fact name -> strict-greater max-cosine threshold.
        let thresholds: [String: Double]
        /// Fixed centering mean, dim must match the model (3072). Nil = no centering.
        let mean: [Double]?
    }

    private let queue = DispatchQueue(label: "ai.ecoinference.needle")
    private let stateLock = NSLock()
    private var isReady = false

    // Guarded by queue: mutated only in blocks running on it.
    private var dim = 0
    private var config: ExemplarConfig?
    private var mean: [Float]?
    private var exemplars: [String: [[Float]]] = [:]

    private init() {}

    /// Fire-and-forget engine + bundled-config startup; call once at launch.
    func start() {
        queue.async { self.initializeOnQueue() }
    }

    /// Version-gated live update (Remote Config `router_exemplars`).
    func applyConfigIfNewer(_ newConfig: ExemplarConfig) {
        queue.async {
            if newConfig.version > (self.config?.version ?? 0) {
                self.applyConfigOnQueue(newConfig)
            }
        }
    }

    /// Semantic facts for one prompt, or nil when the embedder is unavailable
    /// or anything goes wrong — callers keep their keyword-derived facts.
    /// ~8-10 ms per prompt on M-series Macs; the Lenovo TB336FU measurement
    /// is still owed (ROUTING.md §4 item 2).
    func semanticFacts(prompt: String) -> [String: FactValue]? {
        stateLock.lock()
        let ready = isReady
        stateLock.unlock()
        guard ready else { return nil }
        return queue.sync {
            guard let config else { return nil }
            guard let vec = embedOnQueue(ExemplarScorer.normalize(prompt)) else { return nil }
            return ExemplarScorer.semanticFacts(
                promptVec: vec, mean: mean, exemplars: exemplars, thresholds: config.thresholds
            )
        }
    }

    // MARK: - queue-thread internals

    private func initializeOnQueue() {
        guard let url = Bundle.main.url(forResource: "needle3", withExtension: "cact"),
              let cact = try? Data(contentsOf: url) else {
            needleLog.info("weights not bundled (run fetch_needle.sh) — router stays on keyword facts")
            return
        }
        let d = NeedleBridge.loadModel(from: cact)
        guard d > 0 else {
            needleLog.info("engine unavailable: \(NeedleBridge.lastErrorMessage())")
            return
        }
        dim = Int(d)
        stateLock.lock(); isReady = true; stateLock.unlock()
        needleLog.info("engine ready, dim=\(self.dim)")
        if let next = config ?? loadBundledConfig() {
            applyConfigOnQueue(next)
        }
    }

    private func loadBundledConfig() -> ExemplarConfig? {
        guard let url = Bundle.main.url(forResource: "default_router_exemplars", withExtension: "json"),
              let data = try? Data(contentsOf: url) else {
            needleLog.warning("bundled exemplar config missing")
            return nil
        }
        do {
            return try JSONDecoder().decode(ExemplarConfig.self, from: data)
        } catch {
            needleLog.warning("bundled exemplar config unreadable: \(error.localizedDescription)")
            return nil
        }
    }

    private func applyConfigOnQueue(_ newConfig: ExemplarConfig) {
        config = newConfig
        stateLock.lock(); let ready = isReady; stateLock.unlock()
        guard ready else { return }   // embedded later, at the end of initializeOnQueue
        var m: [Float]? = nil
        if let list = newConfig.mean {
            if list.count != dim {
                needleLog.warning("config v\(newConfig.version): mean dim \(list.count) != \(self.dim), ignoring mean")
            } else {
                m = list.map { Float($0) }
            }
        }
        var vectors: [String: [[Float]]] = [:]
        for (cat, texts) in newConfig.categories {
            vectors[cat] = texts.compactMap { raw in
                guard let v = embedOnQueue(ExemplarScorer.normalize(raw)) else { return nil }
                return m.map { ExemplarScorer.center(v, mean: $0) } ?? v
            }
        }
        mean = m
        exemplars = vectors
        needleLog.info("exemplar config v\(newConfig.version) applied: \(vectors.values.map(\.count).reduce(0, +)) vectors")
    }

    private func embedOnQueue(_ text: String) -> [Float]? {
        var out = [Float](repeating: 0, count: dim)
        let r = out.withUnsafeMutableBufferPointer { buf in
            NeedleBridge.embedText(text, into: buf.baseAddress, capacity: dim)
        }
        guard r >= 0 else {
            needleLog.warning("embed failed: \(NeedleBridge.lastErrorMessage())")
            return nil
        }
        return out
    }
}

File: AIAndroid/app/src/main/kotlin/ai/ecoinference/app/router/NeedleEmbedder.kt
package ai.ecoinference.app.router

import android.content.Context
import android.util.Log
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.util.concurrent.Executors

/**
 * Cactus Needle 3 as a fixed feature extractor for the router's semantic
 * facts (docs/ROUTING.md §4 — adopted design). The model ships once; ALL
 * tunable intelligence (exemplar texts, thresholds, centering mean) lives in
 * [ExemplarConfig] JSON — bundled default plus live updates via Firebase
 * Remote Config key `router_exemplars`, same version-gated pattern as
 * `router_rules`. Mirrors iOS NeedleEmbedder.swift.
 *
 * Failure contract: every failure path leaves [semanticFacts] returning null
 * and the router falls back to keyword facts. Routing must never block or
 * crash chat. Engine unavailable (stub ABI, missing weights, init error) is a
 * normal state, not an error.
 *
 * Threading: the needle C API is one process-global, NOT thread-safe model.
 * Every native call runs on [executor]; callers may invoke [semanticFacts]
 * from any thread except the executor itself (it blocks on submit().get()).
 */
object NeedleEmbedder {

    private const val TAG = "NeedleEmbedder"

    @Serializable
    data class ExemplarConfig(
        val version: Int,
        /** fact name -> exemplar texts (embedded once per config load). */
        val categories: Map<String, List<String>>,
        /** fact name -> strict-greater max-cosine threshold. */
        val thresholds: Map<String, Double>,
        /** Fixed centering mean, dim must match the model (3072). Null = no centering. */
        val mean: List<Double>? = null,
    )

    private val json = Json { ignoreUnknownKeys = true }
    private val executor = Executors.newSingleThreadExecutor { r ->
        Thread(r, "needle-embedder").apply { isDaemon = true }
    }

    @Volatile private var ready = false
    @Volatile private var dim = 0

    // Guarded by executor: mutated only in tasks running on it.
    private var config: ExemplarConfig? = null
    private var mean: FloatArray? = null
    private var exemplars: Map<String, List<FloatArray>> = emptyMap()

    private external fun nativeLoad(cact: ByteArray): Int
    private external fun nativeEmbed(text: String, out: FloatArray): Int
    private external fun nativeLastError(): String
    @Suppress("unused") private external fun nativeReset()

    /** Fire-and-forget engine + bundled-config startup; call once from Application.onCreate. */
    fun start(context: Context) {
        val appContext = context.applicationContext
        executor.execute {
            try {
                initOnExecutor(appContext)
            } catch (t: Throwable) {
                // UnsatisfiedLinkError (stub ABI), missing asset, native failure —
                // all mean "embedder unavailable", and keyword facts carry on.
                Log.i(TAG, "embedder unavailable: ${t.message}")
            }
        }
    }

    /** Version-gated live update (Remote Config `router_exemplars`). */
    fun applyConfigIfNewer(newConfig: ExemplarConfig) {
        executor.execute {
            if (newConfig.version > (config?.version ?: 0)) {
                applyConfigOnExecutor(newConfig)
            }
        }
    }

    /**
     * Semantic facts for one prompt, or null when the embedder is unavailable
     * or anything goes wrong — callers keep their keyword-derived facts.
     * ~8-10 ms per prompt on M-series Macs; the Lenovo TB336FU measurement is
     * still owed (ROUTING.md §4 item 2).
     */
    fun semanticFacts(prompt: String): Map<String, FactValue>? {
        if (!ready) return null
        return try {
            executor.submit<Map<String, FactValue>?> {
                val c = config ?: return@submit null
                val vec = embedOnExecutor(ExemplarScorer.normalize(prompt)) ?: return@submit null
                ExemplarScorer.semanticFacts(vec, mean, exemplars, c.thresholds)
            }.get()
        } catch (t: Throwable) {
            Log.w(TAG, "semanticFacts failed, failing open: ${t.message}")
            null
        }
    }

    // ── executor-thread internals ──────────────────────────────────────────

    private fun initOnExecutor(context: Context) {
        System.loadLibrary("needle_jni")
        val cact = context.assets.open("needle3.cact").use { it.readBytes() }
        val d = nativeLoad(cact)
        if (d <= 0) {
            Log.i(TAG, "nativeLoad failed ($d): ${nativeLastError()}")
            return
        }
        dim = d
        ready = true
        Log.i(TAG, "engine ready, dim=$dim")
        val next = config ?: loadBundledConfig(context) ?: return
        applyConfigOnExecutor(next)
    }

    private fun loadBundledConfig(context: Context): ExemplarConfig? = try {
        val raw = context.assets.open("default_router_exemplars.json")
            .bufferedReader().use { it.readText() }
        json.decodeFromString(ExemplarConfig.serializer(), raw)
    } catch (e: Exception) {
        Log.w(TAG, "bundled exemplar config unreadable: ${e.message}")
        null
    }

    private fun applyConfigOnExecutor(newConfig: ExemplarConfig) {
        config = newConfig
        if (!ready) return  // embedded later, at the end of initOnExecutor
        val m = newConfig.mean?.let { list ->
            if (list.size != dim) {
                Log.w(TAG, "config v${newConfig.version}: mean dim ${list.size} != $dim, ignoring mean")
                null
            } else FloatArray(dim) { list[it].toFloat() }
        }
        val vectors = newConfig.categories.mapValues { (_, texts) ->
            texts.mapNotNull { raw ->
                embedOnExecutor(ExemplarScorer.normalize(raw))?.let { v ->
                    if (m != null) ExemplarScorer.center(v, m) else v
                }
            }
        }
        mean = m
        exemplars = vectors
        Log.i(TAG, "exemplar config v${newConfig.version} applied: " +
            "${vectors.values.sumOf { it.size }} vectors")
    }

    private fun embedOnExecutor(text: String): FloatArray? {
        val out = FloatArray(dim)
        return if (nativeEmbed(text, out) < 0) {
            Log.w(TAG, "embed failed: ${nativeLastError()}")
            null
        } else out
    }
}
